packages/flutter/lib/src/services/text_input.dart

```diff
@@ -671,6 +686,381 @@ class RawFloatingCursorPoint {
   final FloatingCursorDragState state;
 }
 
+/// A way to disambiguate a [TextEditingDelta] when a delta generated for an insertion
+/// and a deletion both contain collapsed [TextEditingDelta.deltaRange]'s.
+enum TextEditingDeltaType {
+  /// {@template flutter.services.TextEditingDeltaInsertion}
+  /// The delta is inserting a single/or contigous sequence of characters.
+  /// {@endtemplate}
+  insertion,
+
+  /// {@template flutter.services.TextEditingDeltaDeletion}
+  /// The delta is deleting a single/or contiguous sequence of characters.
+  /// {@endtemplate}
+  deletion,
+
+  /// {@template flutter.services.TextEditingDeltaReplacement}
+  /// The delta is replacing a range of characters with a new sequence of text.
+  /// The range that is being replaced can either grow or shrink based on the
+  /// given replacement text.
+  /// A replacement can occur in cases such as auto-correct, suggestions, and
+  /// when a selection is replaced by a single character.
+  /// {@endtemplate}
+  replacement,
```

Comment:
Isn't `insertion` equivalent to a `replacement` on a collapsed range? Adding a hardcoded type like this for something that is semantically equivalent to existing API may result in redundancy, though this does make a bit more sense to laypeople

Reply:
Same goes for deletion, is it not equivalent to a replacement with an emptystring? Do we have a way to disambiguate between an emptystring replacement vs a deletion, collapsed replacement vs insertion?

Reply:
I'd like to offer a couple thoughts on this. One of them is specific, the other is more generic to the overall approach.

> something that is semantically equivalent to existing API

I think this statement is a bit inaccurate. The changes are "technically equivalent", they are not "semantically equivalent". A user who "inserts" content is not a user who is "replacing" content. From a user/domain perspective, these are distinct intentions, typically utilized for different use-cases. There could be any number of app-level reactions that care about the semantic distinction.

Second, I'd like to repeat a point that I've made elsewhere in this project, but it's on direct display in this particular question/recommendation. This proposed delta API, along with this recommendation to combine event types, are both predicated on the idea of a "least common denominator" approach to fundamental text input. It's a reduction of nuanced platform behaviors to what the Flutter team deems worthy of Flutter apps. Consider that the details that are made available at this bottleneck point within the framework are the only details that any app developer can ever gain access to (without maintaining a permanent fork of the engine).

Historically, Flutter has claimed to oppose the "least common denominator" problem that tends to arise in cross-platform toolkits, but this problem began as a result of an artificially limited API, and we seem to be continuing that trend with this delta PR. Flutter suggests that we embrace the yak shave, but this project is still lacking a deep audit of text-based APIs on the 6 major platforms. How do we know what we're leaving out? Flutter believes in aggressive composition, but rather than port each platforms text APIs and then compose them into a unified API, this approach creates a unified API monolith, from which app developers cannot escape (without maintaining a permanent fork of the engine).

I recommend retaining absolutely all text-related information within the fundamental text APIs, both with regard to this specific area, as well as the overall project, in general. This recommendation is based on the fact that Flutter is a UI toolkit, and text interaction is an absolutely fundamental, universal UI interaction. This is the core of what Flutter exists to do. This is not an area to "simplify" platform APIs. App developers **need** to know what the platform is doing with text.

Reply:
I agree we should definitely avoid a least common denominator approach, which is definitely not the intention here.  The engine PRs have changed a little bit.  Let us know if there is any information from the platform that is being thrown away.